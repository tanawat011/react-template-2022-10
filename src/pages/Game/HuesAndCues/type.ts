/* istanbul ignore file */
import type { DocumentData, QueryDocumentSnapshot } from 'firebase/firestore'

export type Player = {
  id: string
  name: string
}

export type Room = {
  id: string
  password: string
  hintChoice: string[]
  hintSelected: string
  isStarted: boolean
  isSubmitResult: boolean
  totalRound: number
}

export type RawRoomPlayer = {
  refPlayer: QueryDocumentSnapshot<DocumentData>
  color: string
  score: number
  allSelected: string[]
  isOwner: boolean
  isTurn: boolean
  isHinter: boolean
  totalTurn: number
  seq: number
}

export type RoomPlayer = Omit<RawRoomPlayer, 'refPlayer'> & {
  player: Player
}
