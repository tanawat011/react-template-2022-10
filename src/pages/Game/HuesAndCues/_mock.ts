import type { RoomPlayer } from './_type'

const PLAYER_BASE: RoomPlayer = {
  cells: [],
  uid: '1',
  name: 'Player 1',
  score: 0,
  color: '',
  number: 1,
}

const HINTER: RoomPlayer = {
  uid: 'p1',
  name: 'p1',
  number: 1,
  color: 'bg-[#0758a5]',
  score: 0,
  hinter: true,
  cells: [],
}

const HINTER_HAS_CHOICE: RoomPlayer = {
  ...HINTER,
  choice: ['A1', 'A2', 'A3', 'A4'],
}

const HINTER_HAS_CHOICE_SELECT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE,
  result: 'A1',
}

const HINTER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT,
  isSubmitResult: true,
}

const HINTER_IS_OWNER: RoomPlayer = {
  ...HINTER,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE: RoomPlayer = {
  ...HINTER_HAS_CHOICE,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE_SELECT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT,
  owner: true,
}

const PLAYER: RoomPlayer = {
  uid: 'P2',
  name: 'p2',
  number: 2,
  color: 'bg-[#117822]',
  score: 0,
  cells: [],
}

const PLAYER_MY_TURN: RoomPlayer = {
  ...PLAYER,
  isMyTurn: true,
}

const PLAYER_MY_TURN_SELECT_CELL: RoomPlayer = {
  ...PLAYER_MY_TURN,
  cells: ['A1'],
}

const result = 'F4'
const choice = ['F4', 'A9', 'P27', 'E15']

export const MOCK: {
  ONE_PLAYER: RoomPlayer[]
  TWO_PLAYER: RoomPlayer[]
  TWO_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER: RoomPlayer[]
  THREE_PLAYER_SECOND_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_TURN: RoomPlayer[]
  HINTER_NO_CHOICE: RoomPlayer[]
  HINTER_NO_SELECT_CHOICE: RoomPlayer[]
  HINTER_SELECTED_CHOICE: RoomPlayer[]
  PLAYER_NO_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: RoomPlayer[]
} = {
  ONE_PLAYER: [
    {
      ...PLAYER_BASE,
      cells: ['D17'],
      number: 1,
      hinter: true,
      isMyTurn: true,
      color: 'bg-[#0758a5]',
    },
  ],
  TWO_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
  ],
  TWO_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
  ],
  THREE_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
      owner: true,
      score: 1,
    },
    {
      ...PLAYER_BASE,
      uid: '2',
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      uid: '3',
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_TURN: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      isMyTurn: true,
    },
  ],
  THREE_PLAYER_SECOND_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      hinter: true,
    },
  ],
  HINTER_NO_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_NO_SELECT_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_SELECTED_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
      result,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_NO_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      color: 'bg-[#117822]',
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR_START_GAME: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      result: 'F4',
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      cells: ['F4'],
    },
    {
      ...PLAYER_BASE,
      number: 3,
      cells: ['F5', 'F6'],
    },
  ],
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      isSubmitResult: true,
      countRound: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
}
