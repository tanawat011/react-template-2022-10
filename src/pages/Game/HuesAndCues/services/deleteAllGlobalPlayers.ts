import type { RoomPlayer } from '../type'

import { deleteDocument } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../constants'

export const deleteAllGlobalPlayers = async (players: RoomPlayer[]) => {
  const playerPath = FIRESTORE_PATH.DOC_PLAYER

  await Promise.all(
    players.map(async ({ player: { id } }) => {
      const path = playerPath + `/${id}`

      await deleteDocument(path)
    }),
  )
}
