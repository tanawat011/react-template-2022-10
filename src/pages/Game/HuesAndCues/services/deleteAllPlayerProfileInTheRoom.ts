import type { RoomPlayer } from '../_type'

import { getFirestore } from 'firebase/firestore'

import { deleteDocument } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../_constants'

export const deleteAllPlayerProfileInTheRoom = async (players: RoomPlayer[]) => {
  const db = getFirestore()

  await Promise.all(
    players.map(async ({ uid }) => {
      const collectionPath = `${FIRESTORE_PATH.PLAYERS_COLLECTION}/${uid}`

      await deleteDocument(db, collectionPath)
    }),
  )
}
