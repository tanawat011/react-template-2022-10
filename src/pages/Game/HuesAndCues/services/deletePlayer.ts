import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteCollection } from 'helpers/firebase'

export const deletePlayer = async (playerId: string) => {
  const path = FIRESTORE_PATH.DOC_PLAYER + `/${playerId}`

  await deleteCollection(path)
}
