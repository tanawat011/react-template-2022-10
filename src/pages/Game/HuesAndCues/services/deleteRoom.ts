import { getFirestore } from 'firebase/firestore'

import { deleteCollection } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../_constants'

export const deleteRoom = async (roomId: string) => {
  const db = getFirestore()
  const collectionPath = `${FIRESTORE_PATH.ROOMS_DOCUMENT}/${roomId}`

  await deleteCollection(db, collectionPath)
}
