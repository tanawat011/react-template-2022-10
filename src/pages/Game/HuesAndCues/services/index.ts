export * from './deleteAllPlayerProfileInTheRoom'
export * from './deleteRoom'
export * from './getAllPlayersInTheRoom'
export * from './getPlayerInTheRoom'
export * from './getPlayerProfile'
export * from './setPlayerProfile'
export * from './setPlayerInTheRoom'
export * from './subscribeRoom'

export * from './navigate'
