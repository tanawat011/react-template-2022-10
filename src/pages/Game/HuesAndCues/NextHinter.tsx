import type { RoomPlayer } from './_type'

import { Button } from 'components/Button'

import { setPlayerInTheRoom } from './services'

type Prop = {
  roomId: string
  playersInRoom: RoomPlayer[]
  currentPlayer: RoomPlayer
  updateCurrentHinter: (player: RoomPlayer) => Promise<void>
}

export const NextHinter: React.FC<Prop> = ({
  roomId,
  playersInRoom,
  currentPlayer,
  updateCurrentHinter,
}) => {
  const handlerInCase2PlayersInRoom = async () => {
    const currentHinter: RoomPlayer = {
      ...currentPlayer,
      choice: [],
      countRound: 0,
      countTurn: 0,
      hinter: false,
      isSubmitResult: false,
      result: '',
      isMyTurn: true,
    }

    await updateCurrentHinter(currentHinter)

    const nextHinter = playersInRoom.find((p) => !p.hinter)

    if (nextHinter) {
      const _nextHinter: RoomPlayer = {
        ...nextHinter,
        cells: [],
        countRound: 0,
        countTurn: 0,
        hinter: true,
        isMyTurn: false,
        isStartGame: true,
        isSubmitResult: false,
      }

      await setPlayerInTheRoom(roomId, _nextHinter)
    }
  }

  const handleUpdateCurrentHinter = async () => {
    const currentHinter: RoomPlayer = {
      ...currentPlayer,
      choice: [],
      countRound: 0,
      countTurn: 0,
      hinter: false,
      isSubmitResult: false,
      result: '',
    }

    await updateCurrentHinter(currentHinter)

    return currentHinter
  }

  const handleUpdateNextHinter = async (
    currentHinter: RoomPlayer,
    playersInRoomSort: RoomPlayer[],
  ) => {
    let nextNumber = currentHinter.number
    let nextHinter = playersInRoomSort[nextNumber]

    // ** In case the last player in the room
    if (!nextHinter) {
      nextNumber = 0
      nextHinter = playersInRoomSort[nextNumber]
    }

    // ** In case the next player is the hinter
    if (nextHinter && nextHinter.hinter) {
      nextNumber += 1
      nextHinter = playersInRoomSort[nextNumber] || playersInRoomSort[0]
    }

    const _nextHinter: RoomPlayer = {
      ...nextHinter,
      cells: [],
      choice: [],
      countRound: 0,
      countTurn: 0,
      hinter: true,
      isSubmitResult: false,
      result: '',
      isMyTurn: false,
      isStartGame: true,
    }

    await setPlayerInTheRoom(roomId, _nextHinter)

    return nextNumber + 1
  }

  const handleNextHinter = async () => {
    try {
      // * In case has 2 players in the room
      if (playersInRoom.length === 2) {
        await handlerInCase2PlayersInRoom()

        return
      }

      // * Update Current Hinter
      const currentHinter = await handleUpdateCurrentHinter()

      // * Update Next hinter
      const playersInRoomSort = playersInRoom.sort((a, b) => a.number - b.number)
      let nextNumber = await handleUpdateNextHinter(currentHinter, playersInRoomSort)

      // * Update Next Turn
      let nextPlayerTurn = playersInRoomSort[nextNumber]

      if (!nextPlayerTurn) {
        nextNumber = 0
        nextPlayerTurn = playersInRoomSort[nextNumber]
      }

      const _nextPlayerTurn: RoomPlayer = {
        ...nextPlayerTurn,
        cells: [],
        choice: [],
        countRound: 0,
        countTurn: 0,
        hinter: false,
        isSubmitResult: false,
        result: '',
        isMyTurn: true,
        isStartGame: true,
      }

      await setPlayerInTheRoom(roomId, _nextPlayerTurn)
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log('error', error)
    }
  }

  const isNotStartGame = !currentPlayer?.isStartGame
  const isNotSubmitResult = !currentPlayer?.isSubmitResult

  return (
    <Button disabled={isNotStartGame || isNotSubmitResult} onClick={handleNextHinter}>
      Next Hinter
    </Button>
  )
}
