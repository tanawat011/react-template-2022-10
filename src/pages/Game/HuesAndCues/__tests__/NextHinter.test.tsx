import type { RenderResult } from '@testing-library/react'

import { act, fireEvent, screen, waitFor } from '@testing-library/react'
import { mocked } from 'jest-mock'

import { renderWithProviders } from 'helpers/test'

import { NextHinter } from '../NextHinter'
import { MOCK } from '../_mock'
import { setPlayerInTheRoom } from '../services'

jest.mock('components/Button', () => ({
  Button: (props: React.HTMLAttributes<never>) => {
    return (
      <div {...props} data-testid={'button'}>
        {props.children}
      </div>
    )
  },
}))

jest.mock('../services', () => ({
  setPlayerInTheRoom: jest.fn(),
}))

describe('<NextHinter />', () => {
  test('match snapshot', () => {
    const updateCurrentPlayer = jest.fn()

    const { asFragment } = renderWithProviders(
      <NextHinter
        roomId='room-id-xxx'
        playersInRoom={MOCK.THREE_PLAYER}
        currentPlayer={MOCK.THREE_PLAYER[0]}
        updateCurrentHinter={updateCurrentPlayer}
      />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('2 players, clicks next hinter', async () => {
    const updateCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <NextHinter
          roomId='room-id-xxx'
          playersInRoom={MOCK.TWO_PLAYER}
          currentPlayer={MOCK.TWO_PLAYER[0]}
          updateCurrentHinter={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('more than 2 players, clicks next hinter, but error', async () => {
    mocked(setPlayerInTheRoom).mockRejectedValue('')
    const updateCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <NextHinter
          roomId='room-id-xxx'
          playersInRoom={MOCK.THREE_PLAYER}
          currentPlayer={MOCK.THREE_PLAYER[0]}
          updateCurrentHinter={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('more than 2 players, clicks next hinter', async () => {
    const updateCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <NextHinter
          roomId='room-id-xxx'
          playersInRoom={MOCK.THREE_PLAYER}
          currentPlayer={MOCK.THREE_PLAYER[0]}
          updateCurrentHinter={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('more than 2 players, clicks next hinter and next turn will turn back to first player', async () => {
    const updateCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <NextHinter
          roomId='room-id-xxx'
          playersInRoom={MOCK.THREE_PLAYER_SECOND_IS_HINTER}
          currentPlayer={MOCK.THREE_PLAYER_SECOND_IS_HINTER[1]}
          updateCurrentHinter={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('more than 2 players, clicks next hinter and next hinter will turn back to first player', async () => {
    const updateCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <NextHinter
          roomId='room-id-xxx'
          playersInRoom={MOCK.THREE_PLAYER_LAST_IS_HINTER}
          currentPlayer={MOCK.THREE_PLAYER_LAST_IS_HINTER[2]}
          updateCurrentHinter={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('more than 2 players, clicks next hinter and next player is hinter', async () => {
    const updateCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <NextHinter
          roomId='room-id-xxx'
          playersInRoom={MOCK.THREE_PLAYER_LAST_IS_HINTER}
          currentPlayer={MOCK.THREE_PLAYER_LAST_IS_HINTER[1]}
          updateCurrentHinter={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('more than 2 players, disabled button', async () => {
    const updateCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <NextHinter
          roomId='room-id-xxx'
          playersInRoom={MOCK.PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT}
          currentPlayer={MOCK.PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT[0]}
          updateCurrentHinter={updateCurrentPlayer}
        />,
      )
    })

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })
})
