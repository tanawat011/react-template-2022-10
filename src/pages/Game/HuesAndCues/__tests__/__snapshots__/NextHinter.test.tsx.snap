// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NextHinter /> match snapshot 1`] = `
<DocumentFragment>
  <button
    data-testid="button"
    disabled=""
  >
    Next Hinter
  </button>
</DocumentFragment>
`;
